### PR TITLE
trace.detrend changes data dtype 

### PR DESCRIPTION
**Pull request by krischer**
Similar to issue #256, I find that, after reading data from a Seismic Unix file, detrending changes the dtype.

For example
```
d = obspy.read('Uy_file_single.su', format='SU', byteorder='<')
print d[0].data.dtype
d[0].detrend()
print d[0].data.dtype
```

produces

```
float32
float64
```

This behavior seems a bit unexpected.  Am I missing something?

Thanks,
Ryan

### CODE REVIEW

**Comment by krischer** [comment]
This issue is now a PR fixing it.

I had a closer look and it turns out we were a bit inconsistent. Most methods behaved as expected but some not - the tested simple detrending was such an operation. For future reference: You most likely want to use `Trace.detrend("linear")` instead of the default `"simple"` detrending. See the doc for details: http://docs.obspy.org/packages/autogen/obspy.core.trace.Trace.detrend.html?highlight=detrend#obspy.core.trace.Trace.detrend

This PR adds a large but probably not fully comprehensive test to test the behaviour for most commonly used methods. It tries to assert the following things:

* `int32` and `int64` arrays are upcasted to `float64` for all methods except trimming and decimating. `float32` cannot accurately represent all `int32` numbers so double precision is necessary.
* `float32` is preserved wherever possible. For filtering and generally frequency domain operations it is upcasted to `float64` as the additional precision might make a difference in these cases.
* Also all routines that use splines always upcast to `float64`. SciPy internally works with double precision and I see no reason to convert it back to single precision.
* Same reasoning holds for our custom C routines which usually only work with double precision.

Ready to merge from my point of view. Please review!

**Comment by rmodrak** [comment]
Hi Lion,  The above approach seems logical and consistent. Now understand more the underlying challenges and the accompanying obspy rationale.  Please let me know if I can help by testing the code or closing the issue.  Thanks for these thoughtful changes.  Ryan

**Comment by krischer** [comment]
This should be pretty good - only our minimum dependency test build raises an error. Probably due to a very old numpy or scipy version. I'll look into it next week.

File: obspy/core/tests/test_trace.py [hunk 1]
```diff
@@ -2203,6 +2203,269 @@ def test_remove_response_plot(self):
...
+            np.float64)
+        self.assertEqual(
+            tr_float32.copy().integrate(method="spline").data.dtype,
+            np.float64)
+        self.assertEqual(
+            tr_float64.copy().integrate(method="spline").data.dtype,
+            np.float64)
+
+        # Simulation is an operation in the spectral domain so double
+        # precision is a lot more accurate so its fine here.
```
Comment (QuLogic) [inline obspy/core/tests/test_trace.py]: it's

File: obspy/core/tests/test_trace.py [hunk 2]
```diff
@@ -2203,6 +2203,269 @@ def test_remove_response_plot(self):
...
+        self.assertEqual(
+            tr_float32.copy().detrend("spline", order=3,
+                                      dspline=100).data.dtype,
+            np.float32)
+        self.assertEqual(
+            tr_float64.copy().detrend("spline", order=3,
+                                      dspline=100).data.dtype,
+            np.float64)
+
+        # Tapering.
```
Comment (QuLogic) [inline obspy/core/tests/test_trace.py]: These four don't have an explanation of what they should do (though it's kind of obvious).

File: obspy/signal/detrend.py [hunk 3]
```diff
@@ -23,11 +23,16 @@ def simple(data):
     point of the trace
 
     :param data: Data to detrend, type numpy.ndarray.
-    :return: Detrended data.
+    :return: Detrended data. Returns the original array which has been
+        modified in-place.
```
Comment (QuLogic) [inline obspy/signal/detrend.py]: `np.require` will make a copy if changing the `dtype`, so this is not always true.

**Comment by QuLogic** [comment]
Needs a rebase and once the comments are addressed, I think it's good to merge.

**Comment by megies** [comment]
There's one fail on the minimum dependency Travis build..

**Comment by krischer** [comment]
Implemented the review points, rebased, and added a workaround for the minimum dependency build (some old scipy version causing issues...). let's see what CI says.

**Comment by krischer** [comment]
The minimum dependency build fails because the old scipy versions does not support integrating splines :-(

What do we do in cases like this? We still have to support the old scipy version as Debian wheezy ships with it. Should we raise a warning that the used scipy version is too old or backport the spline integration?

**Comment by krischer** [comment]
We also really need to improve our test coverage...the current 86% are just not good enough...

**Comment by QuLogic** [comment]
This integration is new and one of several methods, no? In that case, I'd say just raise a warning or whatever instead of backporting. It's different from the filtering case where you basically had no other choice.

**Comment by krischer** [comment]
Rebased + force pushed.

I also added tests for the integration methods and attempted a backport of the `antiderivative()` method to old scipy versions. If it works: its only a couple of lines and we gain feature parity for old scipy versions - debian 7 is still going to be around for a while so it seems worthwhile to me. Let's see what CI has to say.

File: obspy/signal/differentiate_and_integrate.py [hunk 4]
```diff
@@ -45,4 +45,19 @@ def integrate_spline(data, dx, k=3, **kwargs):
     time_array = np.linspace(0, (len(data) - 1) * dx, len(data))
     spline = scipy.interpolate.InterpolatedUnivariateSpline(time_array, data,
                                                             k=k)
+
+    # Backport of the antiderivative() method to old scipy versions.
```
Comment (QuLogic) [inline obspy/signal/differentiate_and_integrate.py]: Add version here so we know when to remove it.

Comment (krischer) [inline obspy/signal/differentiate_and_integrate.py]: Done.

**Comment by krischer** [comment]
Yay :-) The minimum dependency build passes! I'll merge once the others pass as well.

**Pull request merged**